### ISSUE TITLE
Rebuild target does not complete successfully

### ISSUE THREAD

**Issue by mellinoe** [dotnet/corefx#415]
Running build /t:rebuild or msbuild /t:rebuild from the root of the repository does not seem to be working as of today:

````
Build FAILED.

"D:\OSS\corefx\build.proj" (rebuild target) (1) ->
"D:\OSS\corefx\src\dirs.proj" (Build target) (2) ->
  D:\OSS\corefx\src\dirs.proj(21,37): error MSB4057: The target "EnsureDependencies" does not exist in the project.
````

**Comment by weshaggard** [comment]
This was broken by @mmitche commit https://github.com/dotnet/corefx/commit/9ce75073b1a9bfe52705ecbd8aa2e6127fa971de#diff-3ad318a39efd856540e8cf813f20ec09. 

@mmitche why did you need to move _RestoreBuildTools to be an InitialTargets? This causes issues with rebuild because we download the build tools and then clean them and then try to build :) which doesn't go to well.

**Comment by weshaggard** [comment]
I think I figured out what happened here, which is when I changed the traversal project to use BuildAllProjects in order to allow us to plug other things before and after Build. That broke the clean restore tools scenario which used to have a BeforeTarget of Build. To fix that looks like @mmitche switched it to an InitialTargets instead. I have tweaked that fix to just change the BeforeTargets to be set to BuildAllProjects with PR #422. This fixes the rebuild issue but @mmitche  please let me if there is another scenario that is broken.

**Comment by weshaggard** [comment]
This should be fixed by PR #422 which has been merged. Let me know if you still see issues with this.
